File: app/routes/fixed_assets.py
# ============================================================================
# Fixed assets — types, register, depreciation runs, disposal, CSV import,
# and the reconciliation report.
# ============================================================================

from datetime import date
from decimal import Decimal
from typing import Literal, Optional

from fastapi import APIRouter, Depends, File, HTTPException, UploadFile
from pydantic import Field
from app.schemas.common import StrictModel
from sqlalchemy.orm import Session

from app.database import get_db
from app.models.fixed_assets import (
    DepreciationMethod,
    FixedAsset,
    FixedAssetStatus,
    FixedAssetType,
)
from app.services.fixed_assets import (
    acquisition_transactions,
    book_value,
    books_cover,
    check_amounts,
    dispose_asset,
    import_assets_csv,
    next_asset_number,
    post_acquisition,
    run_depreciation,
)
from app.services.bank_register import money_text
from app.services.upload_limits import read_limited

router = APIRouter(prefix="/api/fixed-assets", tags=["fixed_assets"])


# ── Schemas ──────────────────────────────────────────────────────────────


class AssetTypeCreate(StrictModel):
    name: str
    description: Optional[str] = None
    asset_account_id: Optional[int] = None
    accumulated_depreciation_account_id: Optional[int] = None
    depreciation_expense_account_id: Optional[int] = None
    depreciation_method: DepreciationMethod = DepreciationMethod.STRAIGHT_LINE
    effective_life_years: Optional[Decimal] = None
    annual_rate: Optional[Decimal] = None


class AssetTypeUpdate(AssetTypeCreate):
    name: Optional[str] = None
    is_active: Optional[bool] = None
    depreciation_method: Optional[DepreciationMethod] = None


class AssetAcquisition(StrictModel):
    """How the purchase reaches the books (services/fixed_assets.py):
    paid_from a bank or card `account_id` (with a check or `reference`);
    opening_balance for an asset owned before the books began (`as_of` the
    day they began, `accumulated_depreciation` taken by then); a `bill_id`
    or `expense_id` already entered; or in_books (nothing posts)."""

    method: Literal["paid_from", "opening_balance", "bill", "expense", "in_books"]
    account_id: Optional[int] = None
    reference: Optional[str] = Field(None, max_length=100)
    as_of: Optional[date] = None
    accumulated_depreciation: Decimal = Decimal("0")
    bill_id: Optional[int] = None
    expense_id: Optional[int] = None


class AssetCreate(StrictModel):
    name: str
    asset_type_id: int
    purchase_date: date
    purchase_price: Decimal
    salvage_value: Decimal = Decimal("0")
    description: Optional[str] = None
    # The register form always says; an API caller that doesn't gets the
    # old behaviour (nothing posted), as the CSV import does.
    acquisition: Optional[AssetAcquisition] = None


class AssetUpdate(StrictModel):
    name: Optional[str] = None
    asset_type_id: Optional[int] = None
    purchase_date: Optional[date] = None
    purchase_price: Optional[Decimal] = None
    salvage_value: Optional[Decimal] = None
    description: Optional[str] = None


class DepreciationRunRequest(StrictModel):
    run_date: date


class DisposalRequest(StrictModel):
    disposal_date: date
    proceeds: Decimal = Decimal("0")
    deposit_account_id: int


def _type_payload(t: FixedAssetType) -> dict:
    return {
        "id": t.id,
        "name": t.name,
        "description": t.description,
        "asset_account_id": t.asset_account_id,
        "accumulated_depreciation_account_id": t.accumulated_depreciation_account_id,
        "depreciation_expense_account_id": t.depreciation_expense_account_id,
        "depreciation_method": t.depreciation_method.value,
        "effective_life_years": (
            float(t.effective_life_years) if t.effective_life_years else None
        ),
        "annual_rate": float(t.annual_rate) if t.annual_rate else None,
        "is_active": t.is_active,
    }


def _asset_payload(a: FixedAsset, posted: dict | None = None) -> dict:
    return {
        "id": a.id,
        "asset_number": a.asset_number,
        "name": a.name,
        "asset_type_id": a.asset_type_id,
        "asset_type_name": a.asset_type.name if a.asset_type else None,
        "status": a.status.value,
        "purchase_date": a.purchase_date.isoformat(),
        "purchase_price": float(a.purchase_price),
        "salvage_value": float(a.salvage_value or 0),
        "accumulated_depreciation": float(a.accumulated_depreciation or 0),
        "book_value": float(book_value(a)),
        "last_depreciation_date": (
            a.last_depreciation_date.isoformat() if a.last_depreciation_date else None
        ),
        "disposal_date": a.disposal_date.isoformat() if a.disposal_date else None,
        "disposal_proceeds": (
            float(a.disposal_proceeds) if a.disposal_proceeds is not None else None
        ),
        "description": a.description,
        # the purchase is in the books (a posting made for this asset)
        "posted": a.id in (posted if posted is not None else {}),
    }


def _payloads(db: Session, assets: list) -> list[dict]:
    posted = acquisition_transactions(db, [a.id for a in assets])
    return [_asset_payload(a, posted) for a in assets]


# ── Asset types ──────────────────────────────────────────────────────────


@router.get("/types")
def list_types(include_inactive: bool = False, db: Session = Depends(get_db)):
    # Companies created before 2.9.0 have no types at all; give them the
    # same default a new company gets, the first time the page asks.
    if db.query(FixedAssetType.id).first() is None:
        from app.seed.fixed_assets import ensure_default_asset_type

        if ensure_default_asset_type(db) is not None:
            db.commit()
    q = db.query(FixedAssetType)
    if not include_inactive:
        q = q.filter(FixedAssetType.is_active.is_(True))
    return [_type_payload(t) for t in q.order_by(FixedAssetType.name).all()]


@router.post("/types", status_code=201)
def create_type(data: AssetTypeCreate, db: Session = Depends(get_db)):
    if db.query(FixedAssetType).filter(FixedAssetType.name == data.name).first():
        raise HTTPException(status_code=409, detail="Asset type already exists")
    row = FixedAssetType(**data.model_dump())
    db.add(row)
    db.commit()
    db.refresh(row)
    return _type_payload(row)


@router.put("/types/{type_id}")
def update_type(type_id: int, data: AssetTypeUpdate, db: Session = Depends(get_db)):
    row = db.get(FixedAssetType, type_id)
    if not row:
        raise HTTPException(status_code=404, detail="Asset type not found")
    for key, value in data.model_dump(exclude_unset=True).items():
        setattr(row, key, value)
    db.commit()
    db.refresh(row)
    return _type_payload(row)


# ── Assets ───────────────────────────────────────────────────────────────


@router.get("")
def list_assets(include_disposed: bool = True, db: Session = Depends(get_db)):
    q = db.query(FixedAsset)
    if not include_disposed:
        q = q.filter(FixedAsset.status == FixedAssetStatus.REGISTERED)
    return _payloads(db, q.order_by(FixedAsset.asset_number).all())


@router.get("/{asset_id}")
def get_asset(asset_id: int, db: Session = Depends(get_db)):
    asset = db.get(FixedAsset, asset_id)
    if not asset:
        raise HTTPException(status_code=404, detail="Asset not found")
    return _payloads(db, [asset])[0]


@router.post("", status_code=201)
def create_asset(data: AssetCreate, db: Session = Depends(get_db)):
    if not db.get(FixedAssetType, data.asset_type_id):
        raise HTTPException(status_code=404, detail="Asset type not found")
    check_amounts(data.purchase_price, data.salvage_value)
    asset = FixedAsset(
        asset_number=next_asset_number(db),
        **data.model_dump(exclude={"acquisition"}),
    )
    db.add(asset)
    db.flush()
    if data.acquisition is not None:
        # A refusal here leaves nothing behind: the asset row goes with it.
        post_acquisition(db, asset, data.acquisition)
    db.commit()
    db.refresh(asset)
    return _payloads(db, [asset])[0]


@router.post("/{asset_id}/post-purchase")
def post_purchase(asset_id: int, data: AssetAcquisition, db: Session = Depends(get_db)):
    """Put the purchase of an asset registered before it was asked for —
    or registered as already in the books when it wasn't — in the books."""
    asset = db.get(FixedAsset, asset_id)
    if not asset:
        raise HTTPException(status_code=404, detail="Asset not found")
    if asset.status == FixedAssetStatus.DISPOSED:
        raise HTTPException(status_code=400, detail="Disposed assets are read-only")
    if asset.id in acquisition_transactions(db, [asset.id]):
        raise HTTPException(
            status_code=400, detail="This asset's purchase is already in the books."
        )
    if data.method == "in_books":
        raise HTTPException(status_code=400, detail="Choose how it was paid for.")
    atype = asset.asset_type
    if atype and atype.asset_account_id:
        covered, held, _ = books_cover(db, atype.asset_account_id)
        if covered:
            raise HTTPException(
                status_code=400,
                detail=(
                    f"The asset account already holds {money_text(held)}, enough for "
                    "every asset registered to it. Posting this purchase would count "
                    "it twice."
                ),
            )
    post_acquisition(db, asset, data)
    db.commit()
    db.refresh(asset)
    return _payloads(db, [asset])[0]


@router.put("/{asset_id}")
def update_asset(asset_id: int, data: AssetUpdate, db: Session = Depends(get_db)):
    asset = db.get(FixedAsset, asset_id)
    if not asset:
        raise HTTPException(status_code=404, detail="Asset not found")
    if asset.status == FixedAssetStatus.DISPOSED:
        raise HTTPException(status_code=400, detail="Disposed assets are read-only")
    fields = data.model_dump(exclude_unset=True)
    if "purchase_price" in fields or "salvage_value" in fields:
        check_amounts(
            fields.get("purchase_price", asset.purchase_price),
            fields.get("salvage_value", asset.salvage_value),
        )
    posted_fields = {
        key
        for key in ("purchase_price", "purchase_date", "asset_type_id")
        if key in fields and fields[key] != getattr(asset, key)
    }
    if posted_fields and asset.id in acquisition_transactions(db, [asset.id]):
        raise HTTPException(
            status_code=400,
            detail=(
                "This asset's purchase is posted to the books, so its price, "
                "purchase date and type can't change here. Post a correcting "
                "journal entry, or dispose of it and register it again."
            ),
        )
    for key, value in fields.items():
        setattr(asset, key, value)
    db.commit()
    db.refresh(asset)
    return _payloads(db, [asset])[0]


# ── Operations ───────────────────────────────────────────────────────────


@router.post("/run-depreciation")
def depreciation_run(data: DepreciationRunRequest, db: Session = Depends(get_db)):
    return run_depreciation(db, data.run_date)


@router.post("/{asset_id}/dispose")
def dispose(asset_id: int, data: DisposalRequest, db: Session = Depends(get_db)):
    asset = db.get(FixedAsset, asset_id)
    if not asset:
        raise HTTPException(status_code=404, detail="Asset not found")
    return dispose_asset(
        db, asset, data.disposal_date, data.proceeds, data.deposit_account_id
    )


@router.post("/import-csv")
async def import_csv(file: UploadFile = File(...), db: Session = Depends(get_db)):
    content = await read_limited(file, label="Asset CSV")
    try:
        text = content.decode("utf-8-sig")
    except UnicodeDecodeError:
        text = content.decode("latin-1")
    return import_assets_csv(db, text)


@router.get("/reports/reconciliation")
def reconciliation(db: Session = Depends(get_db)):
    """Fixed Asset Reconciliation: register totals per type (cost,
    accumulated, book value) — compare against the mapped GL accounts."""
    rows = []
    for t in db.query(FixedAssetType).order_by(FixedAssetType.name).all():
        assets = [a for a in t.assets if a.status == FixedAssetStatus.REGISTERED]
        cost = sum(Decimal(str(a.purchase_price)) for a in assets)
        accum = sum(Decimal(str(a.accumulated_depreciation or 0)) for a in assets)
        rows.append(
            {
                "asset_type": t.name,
                "asset_count": len(assets),
                "cost": float(cost),
                "accumulated_depreciation": float(accum),
                "book_value": float(cost - accum),
            }
        )
    return {
        "types": rows,
        "total_cost": sum(r["cost"] for r in rows),
        "total_accumulated": sum(r["accumulated_depreciation"] for r in rows),
        "total_book_value": sum(r["book_value"] for r in rows),
    }

File: app/static/js/fixed_assets.js
/**
 * Fixed Assets — register, depreciation runs, disposal, CSV import.
 * Designed from the joelmacklow fork's fixed-assets slice.
 */
const FixedAssetsPage = {
    async render() {
        const assets = await API.get('/fixed-assets');
        return renderListPage({
            title: 'Fixed Assets',
            headerHtml: `<div class="btn-group">
                    <button class="btn btn-primary" onclick="FixedAssetsPage.showAssetForm()">+ Register Asset</button>
                    <button class="btn btn-secondary" onclick="FixedAssetsPage.showTypeForm()">+ Asset Type</button>
                    <button class="btn btn-secondary" onclick="FixedAssetsPage.showDepreciationForm()">Run Depreciation</button>
                    <button class="btn btn-secondary" onclick="FixedAssetsPage.showImportForm()">Import CSV</button>
                </div>`,
            empty: `<p>No fixed assets registered yet.</p>
                <button class="btn btn-primary" onclick="FixedAssetsPage.showAssetForm()" style="margin-top:10px;">+ Register your first asset</button>`,
            columns: [
                { label: 'Asset #', key: 'asset_number' },
                { label: 'Name', key: 'name' },
                { label: 'Type', key: 'asset_type_name' },
                { label: 'Purchased', key: 'purchase_date' },
                { label: 'Status', key: 'status' },
                { label: 'Cost', cls: 'amount', key: 'purchase_price' },
                { label: 'Accum. Depr.', cls: 'amount', key: 'accumulated_depreciation' },
                { label: 'Book Value', cls: 'amount', key: 'book_value' },
                'Actions',
            ],
            sort: { id: 'fixed-assets', column: 'asset_number', direction: 'asc' },
            items: assets,
            row: a => `<tr data-status="${a.status}">
                    <td><strong>${escapeHtml(a.asset_number)}</strong></td>
                    <td>${escapeHtml(a.name)}</td>
                    <td>${escapeHtml(a.asset_type_name || '')}</td>
                    <td>${formatDate(a.purchase_date)}</td>
                    <td>${escapeHtml(a.status)}</td>
                    <td class="amount">${formatCurrency(a.purchase_price)}</td>
                    <td class="amount">${formatCurrency(a.accumulated_depreciation)}</td>
                    <td class="amount">${formatCurrency(a.book_value)}</td>
                    <td class="actions">
                        ${a.status === 'registered' && !a.posted ? `<button class="btn btn-sm btn-secondary" title="This asset's purchase has no posting of its own" onclick="FixedAssetsPage.showPostPurchaseForm(${a.id})">Post purchase…</button>` : ''}
                        ${a.status === 'registered' ? `<button class="btn btn-sm btn-danger" onclick="FixedAssetsPage.showDisposeForm(${a.id})">Dispose</button>` : ''}
                    </td>
                </tr>`,
        });
    },

    async showTypeForm() {
        const accounts = await API.get('/accounts');
        const opts = kind => accounts
            .filter(a => kind ? a.account_type === kind : true)
            .map(a => `<option value="${a.id}">${escapeHtml(a.account_number || '')} ${escapeHtml(a.name)}</option>`)
            .join('');
        openModal('New Asset Type', `
            <form onsubmit="FixedAssetsPage.saveType(event)">
                <div class="form-grid">
                    <div class="form-group"><label>Name *</label>
                        <input name="name" required placeholder="e.g. Computer Equipment"></div>
                    <div class="form-group"><label>Depreciation Method</label>
                        <select name="depreciation_method" onchange="FixedAssetsPage.methodChanged(this)">
                            <option value="straight_line" selected>Straight line</option>
                            <option value="declining_balance">Declining balance</option>
                        </select></div>
                    <div class="form-group"><label id="fa-life-label">Effective Life (years)</label>
                        <input name="effective_life_years" type="number" step="0.5" value="5"></div>
                    <div class="form-group"><label>Annual Rate (e.g. 0.20)</label>
                        <input name="annual_rate" type="number" step="0.0001"></div>
                    <div class="form-group"><label>Fixed Asset Account</label>
                        <select name="asset_account_id"><option value="">--</option>${opts('asset')}</select></div>
                    <div class="form-group"><label>Accumulated Depreciation Account</label>
                        <select name="accumulated_depreciation_account_id"><option value="">--</option>${opts('asset')}</select></div>
                    <div class="form-group"><label>Depreciation Expense Account</label>
                        <select name="depreciation_expense_account_id"><option value="">--</option>${opts('expense')}</select></div>
                </div>
                <div class="form-actions">
                    <button type="button" class="btn btn-secondary" onclick="closeModal()">Cancel</button>
                    <button type="submit" class="btn btn-primary">Create Type</button>
                </div>
            </form>`);
    },

    methodChanged() { /* labels stay visible; both inputs accepted */ },

    async saveType(e) {
        e.preventDefault();
        const form = e.target;
        const num = v => v ? parseFloat(v) : null;
        const intOrNull = v => v ? parseInt(v) : null;
        try {
            await API.post('/fixed-assets/types', {
                name: form.name.value,
                depreciation_method: form.depreciation_method.value,
                effective_life_years: num(form.effective_life_years.value),
                annual_rate: num(form.annual_rate.value),
                asset_account_id: intOrNull(form.asset_account_id.value),
                accumulated_depreciation_account_id: intOrNull(form.accumulated_depreciation_account_id.value),
                depreciation_expense_account_id: intOrNull(form.depreciation_expense_account_id.value),
            });
            toast('Asset type created');
            closeModal();
        } catch (err) { toast(err.message, 'error'); }
    },

    // How the purchase reaches the books. Registering used to post nothing,
    // while depreciation credited accumulated depreciation, so net equipment
    // went negative (exploratory 2.17.3, W-M19). The form asks now.
    async _acquisitionFields(includeInBooks) {
        const [banks, bills, expenses] = await Promise.all([
            API.get('/accounts?bank=1&active_only=true'),
            API.get('/bills'),
            API.get('/expenses'),
        ]);
        const bankOpts = banks.map(a => `<option value="${a.id}">${escapeHtml(a.account_number || '')} - ${escapeHtml(a.name)}</option>`).join('');
        const docs = bills.filter(b => b.status !== 'void' && b.status !== 'draft').slice(0, 100)
            .map(b => `<option value="bill:${b.id}">Bill ${escapeHtml(b.bill_number)} · ${escapeHtml(b.vendor_name || '')} · ${formatDate(b.date)} · ${formatCurrency(b.total)}</option>`)
            .concat(expenses.filter(x => x.status !== 'void').slice(0, 100)
                .map(x => `<option value="expense:${x.id}">Expense · ${escapeHtml(x.payee || '')} · ${formatDate(x.date)} · ${formatCurrency(x.amount)}${x.reference ? ' · ' + escapeHtml(x.reference) : ''}</option>`))
            .join('');
        return `
            <div class="form-group full-width"><label>How was it paid for? *</label>
                <select name="acq_method" onchange="FixedAssetsPage._acqChanged(this)">
                    <option value="paid_from">Paid from a bank or card account</option>
                    <option value="document">On a bill or expense already entered</option>
                    <option value="opening_balance">Owned before these books began</option>
                    ${includeInBooks ? '<option value="in_books">Already in the books (an opening balance or journal entry put it in the asset account)</option>' : ''}
                </select></div>
            <div class="form-group acq acq-paid_from"><label>Paid from *</label>
                <select name="acq_account_id">${bankOpts}</select></div>
            <div class="form-group acq acq-paid_from"><label>Check / ref #</label>
                <input name="acq_reference" maxlength="100"></div>
            <div class="form-group full-width acq acq-document hidden"><label>Bill or expense *</label>
                <select name="acq_document"><option value="">Choose…</option>${docs}</select>
                <div style="font-size:10px; color:var(--gray-500);">Its cost moves from the expense account the bill or expense used into the asset account.</div></div>
            <div class="form-group acq acq-opening_balance hidden"><label>The day your books began</label>
                <input name="acq_as_of" type="date"></div>
            <div class="form-group acq acq-opening_balance hidden"><label>Depreciation already taken by then</label>
                <input name="acq_taken" type="number" step="0.01" min="0" value="0"></div>
            <div class="form-group full-width acq acq-opening_balance hidden" style="font-size:10px; color:var(--gray-500);">
                Posts the cost to the asset account against Opening Balance Equity (3900), less what was already depreciated.</div>
            <div class="form-group full-width acq acq-in_books hidden" style="font-size:10px; color:var(--gray-500);">
                Nothing is posted. The asset account must already hold the cost of every asset registered to it.</div>`;
    },

    _acqChanged(sel) {
        const form = sel.form;
        form.querySelectorAll('.acq').forEach(el => el.classList.toggle('hidden', !el.classList.contains(`acq-${sel.value}`)));
    },

    _acquisitionFromForm(form) {
        const method = form.acq_method.value;
        if (method === 'paid_from') {
            return { method, account_id: parseInt(form.acq_account_id.value, 10) || null, reference: form.acq_reference.value || null };
        }
        if (method === 'document') {
            const [kind, id] = (form.acq_document.value || ':').split(':');
            if (!kind) return null;
            return kind === 'bill' ? { method: 'bill', bill_id: parseInt(id, 10) } : { method: 'expense', expense_id: parseInt(id, 10) };
        }
        if (method === 'opening_balance') {
            return { method, as_of: form.acq_as_of.value || null, accumulated_depreciation: form.acq_taken.value || '0' };
        }
        return { method };
    },

    async showAssetForm() {
        const types = await API.get('/fixed-assets/types');
        if (!types.length) { toast('Create an asset type first', 'error'); return; }
        const typeOpts = types.map(t => `<option value="${t.id}">${escapeHtml(t.name)}</option>`).join('');
        const acquisition = await FixedAssetsPage._acquisitionFields(true);
        openModal('Register Fixed Asset', `
            <form onsubmit="FixedAssetsPage.saveAsset(event)">
                <div class="form-grid">
                    <div class="form-group"><label>Name *</label>
                        <input name="name" required></div>
                    <div class="form-group"><label>Asset Type *</label>
                        <select name="asset_type_id" required>${typeOpts}</select></div>
                    <div class="form-group"><label>Purchase Date *</label>
                        <input name="purchase_date" type="date" required value="${todayISO()}"></div>
                    <div class="form-group"><label>Purchase Price *</label>
                        <input name="purchase_price" type="number" step="0.01" min="0.01" required></div>
                    <div class="form-group"><label>Salvage Value</label>
                        <input name="salvage_value" type="number" step="0.01" min="0" value="0">
                        <div style="font-size:10px; color:var(--gray-500);">No more than the purchase price.</div></div>
                    ${acquisition}
                    <div class="form-group full-width"><label>Description</label>
                        <textarea name="description"></textarea></div>
                </div>
                <div class="form-actions">
                    <button type="button" class="btn btn-secondary" onclick="closeModal()">Cancel</button>
                    <button type="submit" class="btn btn-primary">Register Asset</button>
                </div>
            </form>`);
    },

    async saveAsset(e) {
        e.preventDefault();
        const form = e.target;
        const price = parseFloat(form.purchase_price.value);
        const salvage = parseFloat(form.salvage_value.value) || 0;
        if (salvage > price) { toast(`Salvage value can't be more than the purchase price (${formatCurrency(price)}).`, 'error'); return; }
        const acquisition = FixedAssetsPage._acquisitionFromForm(form);
        if (!acquisition) { toast('Choose the bill or expense the asset was bought on.', 'error'); return; }
        try {
            await API.post('/fixed-assets', {
                name: form.name.value,
                asset_type_id: parseInt(form.asset_type_id.value),
                purchase_date: form.purchase_date.value,
                purchase_price: form.purchase_price.value,
                salvage_value: form.salvage_value.value || '0',
                description: form.description.value || null,
                acquisition,
            });
            toast(acquisition.method === 'in_books' ? 'Asset registered' : 'Asset registered and its purchase posted');
            closeModal();
            App.navigate('#/fixed-assets');
        } catch (err) { toast(err.message, 'error'); }
    },

    async showPostPurchaseForm(assetId) {
        const acquisition = await FixedAssetsPage._acquisitionFields(false);
        openModal('Post the purchase', `
            <form onsubmit="FixedAssetsPage.postPurchase(event, ${assetId})">
                <p style="font-size:11px; margin-bottom:8px;">
                    This asset's purchase has no posting of its own. Post it only if it isn't in the books yet —
                    if an opening balance or a journal entry already put it in the asset account, leave it.
                </p>
                <div class="form-grid">${acquisition}</div>
                <div class="form-actions">
                    <button type="button" class="btn btn-secondary" onclick="closeModal()">Cancel</button>
                    <button type="submit" class="btn btn-primary">Post Purchase</button>
                </div>
            </form>`);
    },

    async postPurchase(e, assetId) {
        e.preventDefault();
        const acquisition = FixedAssetsPage._acquisitionFromForm(e.target);
        if (!acquisition) { toast('Choose the bill or expense the asset was bought on.', 'error'); return; }
        try {
            await API.post(`/fixed-assets/${assetId}/post-purchase`, acquisition);
            toast('Purchase posted');
            closeModal();
            App.navigate('#/fixed-assets');
        } catch (err) { toast(err.message, 'error'); }
    },

    showDepreciationForm() {
        openModal('Run Depreciation', `
            <form onsubmit="FixedAssetsPage.runDepreciation(event)">
                <div class="form-group"><label>Depreciate through *</label>
                    <input name="run_date" type="date" required value="${todayISO()}"></div>
                <div style="font-size:11px; color:var(--text-muted); margin:8px 0;">
                    Posts one journal per asset (DR depreciation expense /
                    CR accumulated depreciation) for the full months elapsed
                    since each asset's last run.
                </div>
                <div class="form-actions">
                    <button type="button" class="btn btn-secondary" onclick="closeModal()">Cancel</button>
                    <button type="submit" class="btn btn-primary">Run</button>
                </div>
            </form>`);
    },

    async runDepreciation(e) {
        e.preventDefault();
        try {
            const result = await API.post('/fixed-assets/run-depreciation', {
                run_date: e.target.run_date.value,
            });
            toast(`Depreciation posted for ${result.posted} assets (${formatCurrency(result.total)}); ${result.skipped} skipped`);
            closeModal();
            App.navigate('#/fixed-assets');
        } catch (err) { toast(err.message, 'error'); }
    },

    async showDisposeForm(assetId) {
        const accounts = await API.get('/accounts?account_type=asset');
        const opts = accounts.map(a => `<option value="${a.id}">${escapeHtml(a.account_number || '')} ${escapeHtml(a.name)}</option>`).join('');
        openModal('Dispose Asset', `
            <form onsubmit="FixedAssetsPage.dispose(event, ${assetId})">
                <div class="form-grid">
                    <div class="form-group"><label>Disposal Date *</label>
                        <input name="disposal_date" type="date" required value="${todayISO()}"></div>
                    <div class="form-group"><label>Proceeds</label>
                        <input name="proceeds" type="number" step="0.01" value="0"></div>
                    <div class="form-group full-width"><label>Deposit Proceeds To *</label>
                        <select name="deposit_account_id" required>${opts}</select></div>
                </div>
                <div class="form-actions">
                    <button type="button" class="btn btn-secondary" onclick="closeModal()">Cancel</button>
                    <button type="submit" class="btn btn-danger">Dispose</button>
                </div>
            </form>`);
    },

    async dispose(e, assetId) {
        e.preventDefault();
        const form = e.target;
        try {
            const result = await API.post(`/fixed-assets/${assetId}/dispose`, {
                disposal_date: form.disposal_date.value,
                proceeds: parseFloat(form.proceeds.value) || 0,
                deposit_account_id: parseInt(form.deposit_account_id.value),
            });
            const gl = result.gain_loss;
            toast(gl === 0 ? 'Asset disposed' : `Asset disposed (${gl > 0 ? 'gain' : 'loss'} ${formatCurrency(Math.abs(gl))})`);
            closeModal();
            App.navigate('#/fixed-assets');
        } catch (err) { toast(err.message, 'error'); }
    },

    showImportForm() {
        openModal('Import Assets from CSV', `
            <form onsubmit="FixedAssetsPage.importCsv(event)">
                <div class="form-group">
                    <label>CSV file (columns: name, asset_type, purchase_date, purchase_price, salvage_value, description)</label>
                    <input type="file" id="fa-import-file" accept=".csv" required>
                </div>
                <div class="form-actions">
                    <button type="button" class="btn btn-secondary" onclick="closeModal()">Cancel</button>
                    <button type="submit" class="btn btn-primary">Import</button>
                </div>
            </form>`);
    },

    async importCsv(e) {
        e.preventDefault();
        const file = $('#fa-import-file').files[0];
        if (!file) return;
        const formData = new FormData();
        formData.append('file', file);
        try {
            const resp = await fetch('/api/fixed-assets/import-csv', { method: 'POST', body: formData });
            if (!resp.ok) throw new Error(await API.responseError(resp, 'Import failed'));
            const data = await resp.json();
            const errs = data.errors.length ? ` (${data.errors.length} rows failed)` : '';
            toast(`Imported ${data.imported} assets${errs}`);
            if (data.errors.length) console.warn('Asset import errors:', data.errors);
            closeModal();
            App.navigate('#/fixed-assets');
        } catch (err) { toast(err.message, 'error'); }
    },
};
